### PR TITLE
Adds an instrument compiler switch

### PR DESCRIPTION
**Pull request by drognanar**

### CODE REVIEW

**Comment by jaredpar** [comment]
Cc @dotnet/roslyn-compiler

**Comment by tannergooding** [comment]
Also tagging @dotnet/testimpact

File: src/Compilers/Core/Portable/PublicAPI.Shipped.txt [hunk 1]
```diff
@@ -425,7 +425,6 @@ Microsoft.CodeAnalysis.Emit.EmitOptions
 Microsoft.CodeAnalysis.Emit.EmitOptions.BaseAddress.get -> ulong
 Microsoft.CodeAnalysis.Emit.EmitOptions.DebugInformationFormat.get -> Microsoft.CodeAnalysis.Emit.DebugInformationFormat
 Microsoft.CodeAnalysis.Emit.EmitOptions.EmitMetadataOnly.get -> bool
-Microsoft.CodeAnalysis.Emit.EmitOptions.EmitOptions(bool metadataOnly = false, Microsoft.CodeAnalysis.Emit.DebugInformationFormat debugInformationFormat = (Microsoft.CodeAnalysis.Emit.DebugInformationFormat)0, string pdbFilePath = null, string outputNameOverride = null, int fileAlignment = 0, ulong baseAddress = 0, bool highEntropyVirtualAddressSpace = false, Microsoft.CodeAnalysis.SubsystemVersion subsystemVersion = default(Microsoft.CodeAnalysis.SubsystemVersion), string runtimeMetadataVersion = null, bool tolerateErrors = false, bool includePrivateMembers = false) -> void
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/PublicAPI.Shipped.txt]: Never delete from the shipped API file.  Instead add a *REMOVED* entry into the unshipped file.

File: src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb [hunk 2]
```diff
@@ -537,6 +538,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                             sdkPaths.AddRange(ParseSeparatedPaths(value))
                             Continue For
 
+                        Case "instrument"
+                            If String.IsNullOrEmpty(value) Then
+                                AddDiagnostic(diagnostics, ERRID.ERR_ArgumentRequired, "instrument", ":<types>")
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb]: Why do the VB and C# error message differ?

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 3]
```diff
@@ -2678,6 +2679,27 @@ public void ParseOut2()
...
+            Assert.Equal("", parsedArgs.EmitOptions.Instrument);
+
+            parsedArgs = DefaultParse(new[] { @"/instrument:", "a.cs" }, _baseDirectory);
+            parsedArgs.Errors.Verify(
+                // error CS2006: Command-line syntax error: Missing '<text>' for 'instrument' option
+                Diagnostic(ErrorCode.ERR_SwitchNeedsString).WithArguments("<text>", "instrument"));
+
+            parsedArgs = DefaultParse(new[] { "/instrument:", "Test.Flag.Name", "a.cs" }, _baseDirectory);
+            parsedArgs.Errors.Verify(
+                // error CS2006: Command-line syntax error: Missing '<attribute>' for 'instrument' option
```
Comment (agocke) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Why does this say `<attribute>`?

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 4]
```diff
@@ -86,6 +91,35 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
         /// </summary>
         public string RuntimeMetadataVersion { get; private set; }
 
+        // 1.0 BACKCOMPAT OVERLOAD -- DO NOT TOUCH
```
Comment (agocke) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Our previous release is `1.2`, not `1.0`.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 5]
```diff
@@ -2678,6 +2679,27 @@ public void ParseOut2()
             Assert.Null(parsedArgs.CompilationOptions.ModuleName);
         }
 
+        [Fact]
+        public void ParseInstrumentTestNames()
```
Comment (agocke) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Is there a corresponding vb test?

File: src/Compilers/Core/Portable/PublicAPI.Shipped.txt (hunk 1 above)
Comment (tannergooding) [inline src/Compilers/Core/Portable/PublicAPI.Shipped.txt]: @jaredpar, that isn't strictly true. The recommended practice when adding a new default parameter to a shipped method (see https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md) is to create a backwards compatible overload that contains all of the shipped parameters and changes the optional parameters to mandatory. The method containing the optional parameters can then have any new parameters appended to the end.

So the delete is correct, but it should have the following addition replacing it:
```
Microsoft.CodeAnalysis.Emit.EmitOptions.EmitOptions(bool metadataOnly, Microsoft.CodeAnalysis.Emit.DebugInformationFormat debugInformationFormat, string pdbFilePath, string outputNameOverride, int fileAlignment, ulong baseAddress, bool highEntropyVirtualAddressSpace, Microsoft.CodeAnalysis.SubsystemVersion subsystemVersion, string runtimeMetadataVersion, bool tolerateErrors, bool includePrivateMembers) -> void
```

Comment (jaredpar) [inline src/Compilers/Core/Portable/PublicAPI.Shipped.txt]: > that isn't strictly true. 

It is in fact strictly true.  We should **never** delete from the shipped file during a release cycle.  Instead we explicitly list the new API as REMOVED in the unshipped file.

Comment (tannergooding) [inline src/Compilers/Core/Portable/PublicAPI.Shipped.txt]: Ah, I missed the note at the very bottom of the page. That should probably get updated as a more explicit instruction

**Comment by gafter** [comment]
Can you please link this PR to the issue that contains the feature specification?

**Comment by amcasey** [comment]
Shouldn't this be going into features/test-impact?

**Comment by drognanar** [comment]
We're sorting out where to put this

**Pull request closed without merging**